src/Features/Core/Completion/Providers/AbstractSymbolCompletionProvider.cs

```diff
@@ -19,6 +19,14 @@ namespace Microsoft.CodeAnalysis.Completion.Providers
 {
     internal abstract partial class AbstractSymbolCompletionProvider : AbstractCompletionProvider
     {
+        // PERF: Many CompletionProviders dervice AbstractSymbolCompletionProvider and therefore
+        // compute identical contexts. This actually shows up on the 2-core typing test.
+        // Cache the most recent document/position/computed SyntaxContext to reduce repeat computation.
+        private static Document cachedDocument;
+        private static int cachedPosition;
+        private static AbstractSyntaxContext cachedContext;
+        private static readonly object cacheGate = new object();
```

Comment:
For linked files cases, you could grab the set of linked documents and keep around caches for each of them. If documents are processed that are in the current set of caches, then just keep using them. If a document is processed that's not in the current set of caches, then clear the cache and re-populate with the new set of documents linked to the requested one (etc.). This solves the "but how many?" problem posed in the description by using exactly how many will be required (but no more) instead of trying to guess an appropriate constant number.